datanode/datanode.go

```diff
@@ -479,25 +553,49 @@ func (d *dataNode) AssignShardSet(shardSet shard.ShardSet) {
 	}
 
 	for _, shardID := range removing {
-		for _, table := range tables {
-			d.logger.With("table", table, "shard", shardID).Info("removing table shard on placement change")
+		for _, table := range factTables {
+			d.logger.With("table", table, "shard", shardID).Info("removing fact table shard on placement change")
 			d.memStore.RemoveTableShard(table, int(shardID))
+			d.diskStore.DeleteTableShard(table, int(shardID))
 		}
 	}
 
 	for _, shard := range adding {
-		for _, table := range tables {
-			d.logger.With("table", table, "shard", shard.ID(), "state", shard.State()).Info("adding table shard on placement change")
-			d.memStore.AddTableShard(table, int(shard.ID()), shard.State() == m3Shard.Initializing)
+		needPeerCopy := shard.State() == m3Shard.Initializing
+		for _, table := range factTables {
+			d.logger.With("table", table, "shard", shard.ID(), "state", shard.State()).Info("adding fact table shard on placement change")
+			d.memStore.AddTableShard(table, int(shard.ID()), needPeerCopy)
 		}
 	}
+
+	// process dimension tables
+	// add dimension tables only if no previously existing shards owned by data node
+	// remove dimension tables only if no shard left for the node
+	if noExistingShards && !noShardLeft {
+		allShardsInitializing := initializingShards > 0 && len(incoming) == initializingShards
+		for _, table := range dimensionTables {
```

Comment:
what is the initializing logic here for dimension tables?

Reply:
dimension tables will be added when first shard is assigned to the data node
and removed when the last shard is removed from the data node